libraries/botbuilder-ai/src/qnaMakerDialog.ts

```diff
@@ -0,0 +1,359 @@
+/**
+ * @module botbuilder-ai
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { Activity, ActivityTypes } from 'botbuilder-core';
+import { WaterfallDialog, Dialog, DialogTurnResult, DialogContext, WaterfallStepContext, DialogTurnStatus, DialogReason } from 'botbuilder-dialogs';
+import { QnAMakerOptions } from './qnamaker-interfaces/qnamakerOptions';
+import { RankerTypes } from './qnamaker-interfaces/rankerTypes';
+import { QnAMaker, QnAMakerResult } from './';
+import { FeedbackRecord, FeedbackRecords, QnAMakerMetadata } from './qnamaker-interfaces';
+import { QnACardBuilder } from './qnaCardBuilder';
+
+/**
+ * QnAMakerDialog response options.
+ */
+export interface QnAMakerDialogResponseOptions {
+    /**
+    * Title for active learning card.
+    */
+    activeLearningCardTitle: string;
+    /**
+    * Text shown for 'no match' option on active learning card.
+    */
+    cardNoMatchText: string;
+    /**
+    * Activity to be sent in the event of no answer found in KB.
+    */
+    noAnswer: Activity;
+    /**
+    * Activity to be sent in the end that the 'no match' option is selected on active learning card.
+    */
+    cardNoMatchResponse: Activity;
+}
+
+/**
+ * Options for QnAMakerDialog.
+ */
+export interface QnAMakerDialogOptions {
+    /**
+    * Options for QnAMaker knowledgebase.
+    */
+    qnaMakerOptions: QnAMakerOptions;
+    /**
+    * QnAMakerDialog response options.
+    */
+    qnaDialogResponseOptions: QnAMakerDialogResponseOptions;
+}
+
+/**
+ * Query QnAMaker knowledgebase using user utterance. The dialog will also present user with appropriate multi-turn prompt or active learning options.
+ */
+export class QnAMakerDialog extends WaterfallDialog {
+    
+    // state and step value key constants
+    private qnAContextData: string = 'previousContextData';
+    private previousQnAId: string = 'previousQnAId';
+    private options: string = 'options';
+    private qnAData: string = 'qnaData';
+    private currentQuery: string = 'currentQuery';
+
+    // Dialog options parameters
+    private defaultCardNoMatchResponse: string = "Thanks for the feedback.";
```

Comment:
>"Thank [](start = 49, length = 6)

A bit of a nit but we try to use single quote ' or back tick ` for all strings. I'd probably go with back tick for these longer strings.

Reply:
Done - thanks for the tip.